### ISSUE TITLE
[Question] Using ROPC flow in Azure AD B2C AcquireTokenByUsernamePassword returning bad request

### ISSUE THREAD

**Issue by chris-skuvault** [AzureAD/microsoft-authentication-library-for-dotnet#1717]
**Which Version of MSAL are you using ?**
MSAL 4.9.0

**Platform**
netcore3.1

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Auth
    * **[ ] Username Password**
    * [ ] Device code flow (browserless)

Other?

**Is this a new or existing app?**
new library i'm writing

**Expected behavior**
When calling AcquireTokenByUsernamePassword then expect a success response back.

**Actual behavior**
Get a response back saying "AADB2C90057: The provided application is not configured to allow the OAuth Implicit flow."
```csharp
 at Microsoft.Identity.Client.OAuth2.OAuth2Client.ThrowServerException(HttpResponse response, RequestContext requestContext)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.CreateResponse[T](HttpResponse response, RequestContext requestContext, Boolean addCorrelationId)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync[T](Uri endPoint, HttpMethod method, RequestContext requestContext, Boolean expectErrorsOn200OK)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.GetTokenAsync(Uri endPoint, RequestContext requestContext)
   at Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpMessageAsync(String tokenEndpoint)
   at Microsoft.Identity.Client.OAuth2.TokenClient.SendTokenRequestAsync(IDictionary`2 additionalBodyParameters, String scopeOverride, String tokenEndpointOverride, CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.SendTokenRequestAsync(String tokenEndpoint, IDictionary`2 additionalBodyParameters, CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.UsernamePasswordRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenByUsernamePasswordParameters usernamePasswordParameters, CancellationToken cancellationToken)
   at SkuVault.Access.AzureB2CTokenService.AcquireTokenByUsernamePasswordAsync(String userName, String password) in C:\Git\sv-ms-graph\SkuVault.Access\SkuVault.Access\AzureB2CTokenService.cs:line 43
   at SkuVault.Access.Tests.AzureB2CTokenServiceTests.Test1() in C:\Git\sv-ms-graph\SkuVault.Access\SkuVault.Access.Tests\AzureB2CTokenServiceTests.cs:line 35
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.BlockUntilCompleted()
   at NUnit.Framework.Internal.MessagePumpStrategy.NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaitable)
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
```

**Additional context/ Logs / Screenshots**
Having issue finding information about this particular error message. I don't have the implict flow configured and am using the Username Password flow so I'm not sure what the error message is referring too. Was wondering if anyone else has seen this behavior. I followed this article for guidlines on using the flow https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-ropc?tabs=applications. 

Client Creation:
```csharp
var clientApp = PublicClientApplicationBuilder.Create(<ClientId>)
.WithB2CAuthority("https://<Tenant>.b2clogin.com/tfp/<Tenant>.onmicrosoft.com/B2C_1_ROPC_AUTH").Build();
```
Where AcquireTokenByUsernamePassword is called
```csharp
_scopes = new List<string> {"https://graph.microsoft.com/User.ReadWrite.All" };
try
{
        result = await _app.AcquireTokenSilent(_scopes,
	accounts.FirstOrDefault()).ExecuteAsync(_token);
}
catch (MsalUiRequiredException ex)
{
	result = await _app.AcquireTokenByUsernamePassword(_scopes, userName, securePassword).ExecuteAsync(_token);
}
```

**Comment by bgavrilMS** [comment]
Looks like a configuration issue or a B2C bug. Needs investigation. @jennyf19 - do you know if this is supposed to work, especially with a graph scope?

**Comment by bgavrilMS** [comment]
PS: thanks for using ROPC for testing purposes and not hopefully not in an app / website :)

**Comment by jmprieur** [comment]
@chris-skuvault. If you are using a B2C authority, you can only request a scope for a B2C Web API? I don't think that B2C can call MSGraph. Adding @mmacy to confirm
cc: @bgavrilMS @jennyf19

**Comment by mmacy** [comment]
@jmprieur I am not sure on this, perhaps @JasSuri or @yoelhor can help out.

**Comment by chris-skuvault** [comment]
@jmprieur I might be missing something but will try and sum up my experience trying to resolve this issue. I would appreciate any additional guidance you can give.

An app registered in azure ad b2c tenant doesn't support the use of the mircosoft graph api permissions set on the registered application attached to the tenant? 

I tried removing the ms graph scope and only supplying the openid and offline_access scope of my registered application's "Application ID URI" i.e. https://<Tenant>.onmicrosoft.com/<AppIDURI>/ openid and https://<Tenant>.onmicrosoft.com/<AppIDURI>/offline_access and and am still getting back the same error response. I'm not sure this is correct because it seems the openid and offline_access scope live off of the https://graph.microsoft.com/ permissions endpoint i.e. https://graph.microsoft.com/offline_access and https://graph.microsoft.com/openid. Do I need to create these two scopes myself as custom scopes in my App Registration then add the permissions to my app registration "Configured permissions"?

How do I make ROPC sign-in flow work in azure ad b2c work properly using AcquireTokenByUsernamePassword call in MSAL? Is this possible using MSAL library? Or do I need to construct the request myself by hand via https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-ropc?tabs=app-reg-preview based on examples of request in the article. 

It looked possible that MSAL could support this given that there are methods like WithB2CAuthority and AcquireTokenByUsernamePassword and in the documentation for this repo https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/AAD-B2C-specifics#resource-owner-password-credentials-ropc-with-b2c it looks like it should be supported. 

The error message mentioned below getting returned from azure is useless in determining anything. I already know i'm not configured to use the OAuth implicit flow I'm trying to use a different flow intentionally so this message makes no sense.

``{"error":"unauthorized_client","error_description":"AADB2C90057:` The provided application is not configured to allow the OAuth Implicit flow.\r\nCorrelation ID: a2c7c5c7-6e8b-4807-a0bc-2822bd640ed7\r\nTimestamp: 2020-03-19 19:21:24Z\r\n"}``

**Comment by JasSuri** [comment]
> An app registered in azure ad b2c tenant doesn't support the use of the mircosoft graph api permissions set on the registered application attached to the tenant?

Azure AD B2C Application registrations **cannot** request any tokens for Microsoft APsI, or Microsoft Services.
Azure AD B2C Application registrations **can**  request any tokens for your own APIs and their corresponding Application Registrations, by passing in their identifierURI and Scope in the Scopes parameter in MSAL..

Unfortunately, our doc has a bug, when using the Application Registration (preview) menu to register an Azure AD B2C Application Registration, you must enable Implicit Flow (access token), for ROPC to work properly. This was being done for the Azure AD B2C application registrations automatically in the old menu. I don't know why this is, but we are looking into it and the Doc fix will go out very shortly.

**Comment by chris-skuvault** [comment]
@JasSuri Thanks this helped me figure it out! I don't actually need my users to interact with graph api directly and instead can just use my app registration permissions. I had already setup a app with a client secret for this purpose. Is there a problem using app registration for both public and client operations?

In my case I'm using the public client to acquire the username and password and the protected client to interact with graph api.

 I didn't seem to need a second app registration as I don't really have a real web api that needs scoped permissions yet. For now I'm just writing a class library to wrap MSAL for business specific use cases. I did create a read and write scope for my application but ended up just registering them to itself.

This approach seems to work for what I'm trying to achieve though it does seem a little strange.

**Comment by jmprieur** [comment]
Closing as this was answered

**Comment by DRAirey1** [comment]
> PS: thanks for using ROPC for testing purposes and not hopefully not in an app / website :)
I keep hearing people say this, but what is the alternative? publicClientApplication has only three overloads: AcquireTokenByUsernamePassword, AcquireTokenInteractive and AcquireTokenWithDeviceCode. If your suggestion is AcquireTokenWithDeviceCode, then I would greatly appreciate a working example along with a description of how we set up the Azure B2C to handle the work flow.

**Comment by bgavrilMS** [comment]
@DRAirey1 - interactive auth via AcquireTokenInteractive is preferred. I don't think `AcquireTokenWithDeviceCode` works with B2C.

**Comment by DRAirey1** [comment]
Right, that's the conclusion I came to. So why are you recommending we use something we can't use? What is your alternative to ROPC for daemon and console applications where you don't have access to a UI?

**Comment by bgavrilMS** [comment]
@DRAirey1 - console applications are perfectly fine to use a browser. Git, Azure CLI, Azure Powershell and many other command line tools do this. The only tricky bit is an operating system like Linux where you do not have UI. This is a difficult problem and I don't think AAD has figure out the best solution here, but generally Linux box without UI is more of a confidential client and you'd be able to use a cert. 

Daemon app are confidential client apps and you can safely deploy a certificate and use AcquireTokenForClient.

**Comment by DRAirey1** [comment]
> @DRAirey1 - console applications are perfectly fine to use a browser. Git, Azure CLI, Azure Powershell and many other command line tools do this.

That sounds great. I can find no working examples of a Windows Console application using B2C for authentication. A link would be greatly appreciated.

**Comment by bgavrilMS** [comment]
Sadly the team did not create b2c samples for console :(

You can adapt an AAD sample for console - see https://github.com/Azure-Samples/ms-identity-dotnet-desktop-tutorial/tree/master/1-Calling-MSGraph and the Maui (Windows) B2C sample - https://github.com/Azure-Samples/ms-identity-dotnetcore-maui/tree/main/MauiAppB2C

**Comment by DRAirey1** [comment]
> Happy to discuss any issues you encounter.
I don't think this works as well as you would hope. I took my working B2C WinUI application and ripped the authentication code out, place the authentication code in a simple console application, and ran it. I got an embedded Web Browser to pop up, but it had nothing in it. Same Public Client Application id, same scope, same URI callback, empty login dialog.

I've decided that this console application (it loads data into the shared data model) is a confidential application as it's function is closer to a daemon than an end-user application. I'm switching to a Confidential Client Application and having better luck.

Thanks for the philosophical justification that allowed me to investigate the Confidential Client App for this feature.
